src/main/java/com/google/cloud/spanner/myadapter/session/SessionState.java

```diff
@@ -29,4 +82,100 @@ public ProtocolStatus getProtocolStatus() {
     this.protocolStatus = protocolStatus;
   }
+
+  Map<String, SystemVariable> getVariableMapForScope(SessionVariableScope scope) {
+    switch (scope) {
+      case GLOBAL:
+        return GLOBAL_SETTINGS;
+      case SESSION:
+        return settings;
+      default:
+        throw unknownParamError("scope");
+    }
+  }
+  /**
+   * Sets the value of the specified setting. The new value will be persisted if the current
+   * transaction is committed. The value will be lost if the transaction is rolled back.
+   */
+  public void set(String name, String value, SessionVariableScope scope) {
+    logger.log(
+        Level.INFO,
+        () -> String.format("Setting system variable %s to %s at scope %s", name, value, scope));
+    Map<String, SystemVariable> variableMap = getVariableMapForScope(scope);
+    internalSet(name.toLowerCase(Locale.ROOT), value, variableMap);
+  }
+
+  private void internalSet(String name, String value, Map<String, SystemVariable> variableMap) {
+    if ("names".equals(name)) {
+      // TODO: Consider handling "SET NAMES" as a separate statement type.
+      handleNames(value);
+      return;
+    }
+    if ("autocommit".equals(name)) {
+      // Autocommit value needs to be converted to an integer as internally autocommit is being
+      // tracked as integer.
+      value = handleAutocommit(value);
+    }
+    SystemVariable variable = variableMap.get(name);
+    if (variable == null) {
+      throw unknownVariableError(name);
+    }
+    variable.setValue(value);
+  }
+
+  private String handleAutocommit(String value) {
+    if (BooleanParser.TRUE_VALUES.contains(value)) {
+      backendConnection.processSetAutocommit();
+      return "1";
+    }
+    if (BooleanParser.FALSE_VALUES.contains(value)) {
+      backendConnection.processUnsetAutocommit();
+      return "0";
+    }
+    throw invalidValueError("autocommit", value);
+  }
+
+  private void handleNames(String value) {
+    logger.log(Level.INFO, () -> String.format("Setting all character sets to %s", value));
+    List<String> charasets =
+        ImmutableList.of(
+            "character_set_client", "character_set_connection", "character_set_results");
+    Map<String, SystemVariable> variableMap = getVariableMapForScope(SessionVariableScope.SESSION);
+    for (String charset : charasets) {
+      SystemVariable variable = variableMap.get(charset);
+      Preconditions.checkNotNull(variable);
+      variable.setValue(value);
+    }
+  }
+
+  /** Returns the current value of the specified setting. */
+  public SystemVariable get(String name, SessionVariableScope scope) {
+    Map<String, SystemVariable> variableMap = getVariableMapForScope(scope);
+    return internalGet(name.toLowerCase(Locale.ROOT), variableMap);
+  }
+
+  private SystemVariable internalGet(String key, Map<String, SystemVariable> variableMap) {
+    SystemVariable variable = variableMap.get(key);
+    if (variable == null) {
+      throw unknownVariableError(key);
```

Comment:
Is this inline with MySql?

Reply:
Yes, although the error message shown was not completely consistent. I'll fix it in the latest commit.